### PR TITLE
Separate diagnostics

### PR DESCRIPTION
**Pull request by DavidM-D**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 1]
```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
...
+noVersions =
+    error "All store items must have versions"
+
+fromUri :: LSP.Uri -> FilePath
+fromUri = fromMaybe noFilePath . uriToFilePath
+
+-- TODO (MK) We canonicalize to make sure that the two files agree on use of
+-- / and \ and other shenanigans.
+-- haskell-lsp also has serious problems with relative file paths
+toUri :: FilePath -> IO LSP.Uri
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: This is a sufficiently bad hack, that I would prefer to inline the definition and add a comment each time we have to do this. Also as mentioned above, using different keys for diagnostics and values seems problematic. `makeAbsolute` might be more appropriate than `canonicalizePath` but doesn’t solve the problem. We should have some tests that test weird combinations of filepaths.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 2]
```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
...
+    concatMap (\(k,v) -> map (fromUri k,) $ getDiagnosticsFromStore v) . Map.toList . getStore
+
+getFileDiagnostics ::
+    FilePath ->
+    Diagnostics stage ->
+    IO [LSP.Diagnostic]
+getFileDiagnostics fp ds = do
+    uri <- toUri fp
+    pure $
+        maybe [] getDiagnosticsFromStore $
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: `findWithDefault [] uri $ getStore ds` works nicely here.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 3]
```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
...
+        (StoreItem _ f) <- Map.lookup uri ds
+        toList <$> Map.lookup (Just $ T.pack $ show stage) f
+
+filterDiagnostics ::
+    (FilePath -> Bool) ->
+    Diagnostics stage ->
+    Diagnostics stage
+filterDiagnostics keep =
+    Diagnostics .
+    Map.filterWithKey (\file _ -> maybe False keep $ uriToFilePath file) .
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Under which circumstances do we expect `uriToFilePath` to fail? When we use `noFilePath`? If so, do  we really want to filter these?

File: daml-foundations/daml-ghc/tests/Bad.daml [hunk 4]
```diff
@@ -5,3 +5,4 @@ daml 1.2
 module Bad where -- Used by test `Import.daml`
 
 data T = T {}
+
```
Comment (cocreature) [inline daml-foundations/daml-ghc/tests/Bad.daml]: Why did you change this?

File: daml-foundations/daml-ghc/tests/ide/warning_tab.daml [hunk 5]
```diff
@@ -11,3 +11,6 @@ module Warning_tab where
 
 a = x
 	where x = 10
+
```
Comment (cocreature) [inline daml-foundations/daml-ghc/tests/ide/warning_tab.daml]: Why the change here?

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 6]
```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
...
+        [(FilePath, [(T.Text, [LSP.Diagnostic])])]
+        -- ^ Source File, Stage Source, Diags
+    storeContents =
+        map (\(uri, StoreItem _ si) ->
+                 (fromMaybe dontKnow $ uriToFilePath uri, getDiags si))
+            $ Map.assocs
+            $ getStore
+            $ removeEmptyStages ds
+
+    dontKnow :: IsString s => s
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: What is the difference between `dontKnow` and `noFilePath`?

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 2 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: I don't think it does looking up by URI returns a `StoreItem` which needs to be unpacked by getDiagnostics from store, I could do this with an empty StoreItem as the default, but I don't think it's clearer

Comment (cocreature) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Oh yeah, sorry missed that.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 7]
```diff
@@ -184,3 +199,170 @@ prettyFileDiagnostics (filePath, diags) =
 getDiagnosticsFromStore :: StoreItem -> [Diagnostic]
 getDiagnosticsFromStore (StoreItem _ diags) =
     toList =<< Map.elems diags
+
+-- | A wrapper around the lsp diagnostics store, the constraint describes how compilation steps
+--   are represented
+newtype Diagnostics stage = Diagnostics {getStore :: DiagnosticStore}
```
[the remark this answers is not in the document]
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: I think `ProjectDiagnostics` makes the most sense

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 6 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Dontknow

**Comment by DavidM-D** [comment]
@cocreature I think we just need to make out paths consistent, but I think it makes sense to do this relative to the project root. Thoughts?

**Comment by cocreature** [comment]
Making paths relative to the project root sounds reasonable (although we should make sure that things also still work properly when we are outside of a project root). I would however suggest that we do this in 3 steps (and 3 PRs):
1. For this PR, leave the filepaths in `Values` unmodified and change `ProjectDiagnostics` to use URIs that correspond exactly to the filepaths used in `Values` so no kind of canonialization or makeAbsolute applied.
2. Add a test suite that checks all the various edge cases, so absolute and relative paths that map to the same file, different path separators, symlinks, …. That test suite should probably test both filepaths at the LSP level and filepaths that go directly into the shake layer. I am fairly sure that the current logic is already broken in various ways that are just hard to trigger so just mark these as expected failures.
3. Then finally, change how filepaths are represented, e.g., by making them relative to the project root and make sure that whatever approach we choose can handle all the edge cases in the test suite.

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs [hunk 8]
```diff
@@ -119,7 +117,7 @@ getIdeGlobalState = getIdeGlobalExtras . shakeExtras
 type Values =
     Map.HashMap FilePath
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]: Given we are now just storing values wouldn't Map (FilePath, Key) be better? The nested structure was only to get diagnostics.

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs [hunk 9]
```diff
@@ -257,26 +258,27 @@ shakeRun IdeState{shakeExtras=ShakeExtras{..}, ..} acts = modifyVar shakeAbort $
 -- | Use the last stale value, if it's ever been computed.
 useStale
     :: IdeRule k v
-    => IdeState -> k -> FilePath -> IO (Maybe v)
-useStale IdeState{shakeExtras=ShakeExtras{state}} k fp =
+    => ShakeExtras -> k -> FilePath -> IO (Maybe v)
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]: Why did this change type? Why not get the ShakeExtras from inside IdeState? I imagine ShakeExtras won't be exposed publicly, but IdeState will.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 10]
```diff
@@ -184,3 +174,155 @@ prettyFileDiagnostics (filePath, diags) =
 getDiagnosticsFromStore :: StoreItem -> [Diagnostic]
 getDiagnosticsFromStore (StoreItem _ diags) =
     toList =<< Map.elems diags
+
+-- | This represents every diagnostic in a LSP project, the stage type variable is
+--   the type of the compiler stages, in this project that is always the Key data
+--   type found in Development.IDE.State.Shake
+newtype ProjectDiagnostics stage = ProjectDiagnostics {getStore :: DiagnosticStore}
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: What's the advantage of this newtype?

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 11]
```diff
@@ -184,3 +174,155 @@ prettyFileDiagnostics (filePath, diags) =
...
+
+    fileErrors :: (FilePath, [(T.Text, [LSP.Diagnostic])]) -> [Doc SyntaxClass]
+    fileErrors (filePath, stages) =
+        [ slabel_ "File:" $ pretty filePath
+        , slabel_ "Errors:" $ vcat $ map prettyStage stages
+        ]
+
+    prettyStage :: (T.Text, [LSP.Diagnostic]) -> Doc SyntaxClass
+    prettyStage (stage,diags) =
+        slabel_ ("Stage: "<>T.unpack stage) $ vcat $ concatMap diagSC diags
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Let's not include the stage labels. They've never been helpful.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 12]
```diff
@@ -184,3 +174,155 @@ prettyFileDiagnostics (filePath, diags) =
...
+    getDiags :: DiagnosticsBySource -> [(T.Text, [LSP.Diagnostic])]
+    getDiags = map (\(ds, diag) -> (fromMaybe (T.pack noFilePath) ds, toList diag)) . Map.assocs
+
+emptyDiagnostics :: ProjectDiagnostics stage
+emptyDiagnostics = ProjectDiagnostics mempty
+
+-- | Sets the diagnostics for a file and compilation step
+--   if you want to clear the diagnostics call this with an empty list
+setStageDiagnostics ::
+  Show stage =>
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: This is odd formatting. Everyone else puts the `->` at the beginning of the line.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 13]
```diff
@@ -184,3 +174,155 @@ prettyFileDiagnostics (filePath, diags) =
...
+        Just (StoreItem Nothing _) ->
+            noVersions
+
+        Just (StoreItem (Just prevTime) bySource) ->
+            case compare posixTime prevTime of
+                EQ -> insert bySource
+                GT -> insert $ Map.map (SL.map tagOutdated)  bySource
+                LT -> bySource
+
+    -- Change the color of outdated diagnostics
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Is this really a good idea. When you type suddenly all the diagnostics will flicker gray and then back to other colors.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 1 above)
Comment (neil-da) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Given horrible hacks, I'd rather we give it a scary name, to toUriHack, and then we can have a single place which describes all the hacks - otherwise one instance will lose the scary comment. Is canonicalize really correct? Does it work with symlinks? Or does it inadvertently chase them, thus giving wrong file errors?

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs (hunk 8 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]: Yeah, that makes sense, we key stuff by file for garbage collection, but that shouldn't be too much trouble

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 11 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: They are pretty useful for debugging, they tell you which shake rule the diagnostics originate from. That being said, they're probably not so useful for the user

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 12 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: I'm quite a fan, it gives you alignment without white space padding, while making it clear that this is a constraint. That being said consistency is probably more important than correctness

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 10 above)
Comment (DavidM-D) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: I'm trying to encapsulate some complexity. There's a lot of wrong ways to add things to the Diagnostics map and we even diverge from the way the haskell-lsp library does it in subtle ways. This explanation probably deserves a comment

File: docs/source/support/release-notes.rst [hunk 14]
```diff
@@ -196,6 +196,8 @@ SQL Extractor
 --------------------
 
 - Making transaction lookups performant so we can handle such requests for large ledgers as well
+- We now color stale IDE diagnostics green to indicate that they are not current
```
Comment (bethaitman) [inline docs/source/support/release-notes.rst]: Looks like you've added this to a previous release, was that deliberate?

Also, nitpick - release notes bullets should end with a full stop.

Comment (DavidM-D) [reply docs/source/support/release-notes.rst]: Good catch! Some of the joys of rebasing

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs [hunk 15]
```diff
@@ -365,46 +369,39 @@ defineEarlyCutoff op = addBuiltinRule noLint noIdentity $ \(Q (key, file)) old m
...
+updateFileDiagnostics fp k ShakeExtras{diagnostics, state, makeRelative} current = do
+    (newDiags, oldDiags) <- liftIO $ do
+        modTime <- join <$> getValues state GetModificationTime fp
+        modifyVar diagnostics $ \old -> do
+            oldDiags <- getFileDiagnostics makeRelative fp old
+            new <- setStageDiagnostics makeRelative fp modTime k current old
+            newDiags <- getFileDiagnostics makeRelative fp new
+            pure (new, (newDiags, oldDiags))
+    when (newDiags /= oldDiags) $
+        sendEvent $ EventFileDiagnostics $ (fp, newDiags)
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]:
```suggestion
        sendEvent $ EventFileDiagnostics (fp, newDiags)
```

File: compiler/haskell-ide-core/test/Development/Ide/Tests.hs [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+main =
+    Tasty.defaultMain $ Tasty.testGroup "Development.IDE.Tests" [fileTests]
+
+-- * File equivalence tests
+makeRelative :: FilePath -> IO FilePath
+makeRelative fp = withProjectRoot ($ fp)
+
+fileTests :: Tasty.TestTree
+fileTests = Tasty.testGroup "File tests"
+    [ Test.case "Paths are always equal when they are URIs" $
```
Comment (neil-da) [inline compiler/haskell-ide-core/test/Development/Ide/Tests.hs]: Going to guess this doesn't compile?

File: daml-foundations/daml-tools/language-server-tests/src/tests/diagnostics.ts [hunk 17]
```diff
@@ -42,7 +42,7 @@ function gatherDiagnostics(diagnostics : rxjs.Observable<DT.protocol.PublishDiag
             DT.tryWriteActualJSON(rootPath, actual);
             done(e, [subscription]);
         }
-    }, 50000);
+    }, 7000);
```
Comment (neil-da) [inline daml-foundations/daml-tools/language-server-tests/src/tests/diagnostics.ts]: What's the reason for this change? What is 7000 here?

File: daml-foundations/daml-tools/language-server-tests/src/tests/issue-reports.ts [hunk 18]
```diff
@@ -6,6 +6,7 @@ import * as fs from 'fs';
 import * as assert from 'assert';
 import * as path from 'path';
 import * as Rx from 'rxjs'
+import { gatherDiagnostics } from './diagnostics';
```
Comment (neil-da) [inline daml-foundations/daml-tools/language-server-tests/src/tests/issue-reports.ts]: Is it important that this change lands in the same PR? I don't see how they are entangled?

File: dev-env/bin/da-ghcid [hunk 19]
```diff
@@ -16,7 +16,7 @@ def main():
             "--data",
             action="store",
             dest="data",
-            choices=["yes","no"],
+            choices=["yes","no", "auto"],
```
Comment (neil-da) [inline dev-env/bin/da-ghcid]: Why is this change happening?

**Comment by DavidM-D** [comment]
Retest this please

File: compiler/haskell-ide-core/BUILD.bazel [hunk 20]
```diff
@@ -80,6 +81,29 @@ da_haskell_binary(
...
     deps = [
         "haskell-ide-core-public",
+        "//daml-assistant:daml-project-config",
         "//libs-haskell/prettyprinter-syntax",
     ],
 ) if not is_windows else None  # Disable on Windows until ghc-paths is fixed upstream
+
+da_haskell_test(
+    name = "test",
+    srcs = glob(["test/Development/Ide/Tests.hs"]),
```
Comment (cocreature) [inline compiler/haskell-ide-core/BUILD.bazel]:
```suggestion
    srcs = glob(["test/Development/IDE/Tests.hs"]),
```

File: daml-assistant/daml-project-config/DAML/Project/Consts.hs [hunk 21]
```diff
@@ -137,15 +139,17 @@ getDamlAssistant = getEnv damlAssistantEnvVar
...
-        Nothing -> act pure
-        Just projectPath -> do
-            projectPath <- canonicalizePath projectPath
-            withCurrentDirectory projectPath $ act $ \f -> do
-                absF <- canonicalizePath (previousCwd </> f)
-                pure (projectPath `makeRelative` absF)
+    let projectPath = fromMaybe previousCwd mbProjectPath
+    projectPath <- canonicalizePath projectPath
+    withCurrentDirectory projectPath $ act $ \f -> do
+        absF <- makeAbsolute (previousCwd </> f)
```
Comment (cocreature) [inline daml-assistant/daml-project-config/DAML/Project/Consts.hs]: Resolving symlinks in `projectPath` but not in `previousCwd </> f` seems like a bad idea. Assuming that there is a reason for this and this is a sane thing to do, we should at least have a comment for this.

File: daml-assistant/daml-project-config/DAML/Project/Consts.hs [hunk 22]
```diff
@@ -137,15 +139,17 @@ getDamlAssistant = getEnv damlAssistantEnvVar
...
+        pure $ projectPath `makeRelative` absF
+
+makeRelativeToRoot :: FilePath -> IO FilePath
```
Comment (cocreature) [inline daml-assistant/daml-project-config/DAML/Project/Consts.hs]: I don’t think this function is a good idea outside of test suites so I’d suggest moving it to the test suite.

File: compiler/haskell-ide-core/BUILD.bazel [hunk 23]
```diff
@@ -64,7 +65,7 @@ da_haskell_library(
 
 da_haskell_binary(
     name = "haskell-ide-core-demo",
-    srcs = glob(["test/**/*.hs"]),
+    srcs = glob(["test/Demo.hs"]),
```
Comment (aherrmann) [inline compiler/haskell-ide-core/BUILD.bazel]:
```suggestion
    srcs = ["test/Demo.hs"],
```
If you're listing sources explicitly, you don't really need to glob anymore. Also if the source file is renamed, then `glob` would return an empty list, which would produce a more confusing error message than listing the file explicitly.

File: compiler/haskell-ide-core/BUILD.bazel (hunk 20 above)
Comment (aherrmann) [inline compiler/haskell-ide-core/BUILD.bazel]:
```suggestion
    srcs = ["test/Development/IDE/Tests.hs"],
```
Same here.

**Comment by DavidM-D** [comment]
Retest this please

**Comment by garyverhaegen-da** [comment]
> Retest this please

FYI the new "retest this please" is `/azp run`. You can also run `git commit --amend` to recreate the current commit with a new commit date and then force-push that.

**Comment by cocreature** [comment]
Merged in a simplified form in #1423. Thanks for all the work you put into this @DavidM-D!
